typescript api package auth refactor blog post (#171)
bluesky-social/bsky-docs: blog/2024-08-12-ts-api-refactor.md

The motivation for these changes is the need to make the `@atproto/api` package compatible with OAuth session management. We don't have OAuth client support "launched" and documented quiet yet, so you can keep using the current app password authentication system. When we do "launch" OAuth support and start encourage it's usage in the near future (see the [OAuth Roadmap](https://github.com/bluesky-social/atproto/discussions/2656)), these changes will make it easier to migrate.

In addition, the redesigned session management system fixes a bug that could cause the session data to become invalid when Agent clones are created (e.g. using `agent.withProxy()`).

## New Features

We've restructured the `XrpcClient` HTTP fetch handler to be specified during the instantiation of the XRPC client, through the constructor, instead of using a default implementation (which was statically defined).

With this refactor, the XRPC client is now more modular and reusable. Session management, retries, cryptographic signing, and other request-specific logic can be implemented in the fetch handler itself rather than by the calling code.

> A new abstract class named `Agent`, has been added to `@atproto/api`. This class will be the base class for all Bluesky agents classes in the `@atproto` ecosystem. It is meant to be extended by implementation that provide session management and fetch handling for the `AtpClient` instances.

Reviewer: ```suggestion
A new abstract class named `Agent`, has been added to `@atproto/api`. This class will be the base class for all Bluesky agents classes in the `@atproto` ecosystem. It is meant to be extended by implementation that provide session management and fetch handling.
```

Not sure what I meant by that. `AtpClient` is kind of an internal implementation detail. I updated the client code & changeset accordingly.

> 

Reviewer: ````suggestion

As you adapt your code to these changes, make sure to use the `Agent` type wherever you expect to receive an agent, and use the `AtpAgent` type (class) only to instantiate your client. The reason for this is to be forward compatible with the OAuth agent implementation that will also extend `Agent`, and not `AtpAgent`.

```ts
import { Agent, AtpAgent } from '@atproto/api'

async function setupAgent(service: string, username: string, password: string): Promise<Agent> {
  const agent = new AtpAgent({
    service,
    persistSession: (evt, session) => {
      // handle session update
    },
  })
  
  await agent.login(username, password)
  
  return agent
}
```

```ts
import { Agent } from '@atproto/api'

async function doStuffWithAgent(agent: Agent, arg: string) {
  return agent.resolveHandle(arg)
}
```

```ts
import { Agent, AtpAgent } from '@atproto/api'

class MyClass {
  agent: Agent

  constructor () {
    this.agent = new AtpAgent()
  }
}
```

````

As you adapt your code to these changes, make sure to use the `Agent` type wherever you expect to receive an agent, and use the `AtpAgent` type (class) only to instantiate your client. The reason for this is to be forward compatible with the OAuth agent implementation that will also extend `Agent`, and not `AtpAgent`.

```ts
import { Agent, AtpAgent } from '@atproto/api'

async function setupAgent(service: string, username: string, password: string): Promise<Agent> {
  const agent = new AtpAgent({
    service,
    persistSession: (evt, session) => {
      // handle session update
    },
  })
  
  await agent.login(username, password)
  
  return agent
}
```

```ts
import { Agent } from '@atproto/api'

async function doStuffWithAgent(agent: Agent, arg: string) {
  return agent.resolveHandle(arg)
}
```

```ts
import { Agent, AtpAgent } from '@atproto/api'

class MyClass {
  agent: Agent

  constructor () {
    this.agent = new AtpAgent()
  }
}
```

## Breaking changes

Most of the changes introduced in this version are backward-compatible. However, there are a couple of breaking changes you should be aware of:

> - Customizing `fetch`: The ability to customize the `fetch: FetchHandler` property of `@atproto/xrpc`'s `Client` and `@atproto/api`'s `AtpAgent` classes has been removed. Previously, the `fetch` property could be set to a function that would be used as the fetch handler for that instance, and was initialized to a default fetch handler. That property is still accessible in a read-only fashion through the `fetchHandler` property. That property can only be set during the instance creation. Attempting to set/get the `fetch` property will now result in an error.

Reviewer 2: ```suggestion
- Customizing `fetch`: The ability to customize the `fetch: FetchHandler` property of `@atproto/xrpc`'s `Client` and `@atproto/api`'s `AtpAgent` classes has been removed. Previously, the `fetch` property could be set to a function that would be used as the fetch handler for that instance, and was initialized to a default fetch handler. That property is still accessible in a read-only fashion through the `fetchHandler` property and can only be set during the instance creation. Attempting to set/get the `fetch` property will now result in an error.
```

- The `fetch()` method, as well as WhatWG compliant `Request` and `Headers` constructors, must be globally available in your environment. Use a polyfill if necessary.
- The `AtpBaseClient` has been removed. The `AtpServiceClient` has been renamed `AtpBaseClient`. Any code using either of these classes will need to be updated.

> - Instead of *wrapping* an `XrpcClient` in its `xrpc` property, the `AtpBaseClient` (formerly `AtpServiceClient`) class, created through `lex-cli`, now *extends* the `XrpcClient` class. This means that a client instance now passes the `instanceof XrpcClient` check. The `xrpc` property now returns the instance itself and has been deprecated.

Reviewer 2: ```suggestion
- Instead of *wrapping* an `XrpcClient` in its `xrpc` property, the `AtpBaseClient` (formerly `AtpServiceClient`) class - created through `lex-cli` - now *extends* the `XrpcClient` class. This means that a client instance now passes the `instanceof XrpcClient` check. The `xrpc` property now returns the instance itself and has been deprecated.
```

The deprecated `ServiceClient` class now extends the new `XrpcClient` class. Because of this, the `fetch` `FetchHandler` can no longer be configured on the `Client` instances (including the default export of the package). If you are not relying on the `fetch` `FetchHandler`, the new changes should have no impact on your code. Beware that the deprecated classes will eventually be removed in a
future version.

Because it's use was completely changed, the `FetchHandler` type was also completely changed. The new `FetchHandler` type is now a function that receives a `url` pathname and a `RequestInit` object and returns a `Promise<Response>`. This function is responsible from making the actual request to the server.

```ts
export type FetchHandler = (
  this: void,
  /**
   * The URL (pathname + query parameters) to make the request to, without the
   * origin. The origin (protocol, hostname, and port) must be added by this
   * {@link FetchHandler}, typically based on authentication or other factors.
   */
  url: string,
  init: RequestInit,
) => Promise<Response>
```

> A noticeable change that was introduced is that the `uri` field of the `ServiceClient` class was _not_ ported to the new `XrpcClient` class. It is now the responsibility of the `FetchHandler` to determine the full URL to make the request to. Same goes for the `headers`, that should now be set through the `FetchHandler` function.

Reviewer 2: ```suggestion
A noticeable change that has been introduced is that the `uri` field of the `ServiceClient` class has _not_ been ported to the new `XrpcClient` class. It is now the responsibility of the `FetchHandler` to determine the full URL to make the request to. The same goes for the `headers`, which should now be set through the `FetchHandler` function.
```

If you _do_ rely on the legacy `Client.fetch` property to perform custom logic upon request, you will need to migrate your code to use the new `XrpcClient` class. The `XrpcClient` class has a similar API to the old `ServiceClient` class, but with a few differences:

- The `Client` + `ServiceClient` duality was removed in favor of a single `XrpcClient` class. This means that:

>   - There no longer exist a centralized lexicon registry. If you need a global lexicon registry, you can maintain one yourself using a `new Lexicons` (from `@atproto/lexicon`).

Reviewer 2: ```suggestion
  - There no longer exists a centralized lexicon registry. If you need a global lexicon registry, you can maintain one yourself using a `new Lexicons` (from `@atproto/lexicon`).
```

  - The `FetchHandler` is no longer a statically defined property of the `Client` class. Instead, it is passed as an argument to the `XrpcClient` constructor.
- The `XrpcClient` constructor now requires a `FetchHandler` function as the first argument, and an optional `Lexicon` instance as the second argument.

> - The `setHeander` and `unsetHeader` methods were not ported to the new `XrpcClient` class. If you need to set or unset headers, you should do so in the `FetchHandler` function provided in the constructor arg.

Reviewer 2: ```suggestion
- The `setHeader` and `unsetHeader` methods were not ported to the new `XrpcClient` class. If you need to set or unset headers, you should do so in the `FetchHandler` function provided in the constructor arg.
```

<table>
<tr>
<td><center>Before</center></td> <td><center>After</center></td>
</tr>
<tr>
<td>

```ts
import client, { defaultFetchHandler } from '@atproto/xrpc'

client.fetch = function (
  httpUri: string,
  httpMethod: string,
  httpHeaders: Headers,
  httpReqBody: unknown,
) {
  // Custom logic here
  return defaultFetchHandler(httpUri, httpMethod, httpHeaders, httpReqBody)
}

client.addLexicon({
  lexicon: 1,
  id: 'io.example.doStuff',
  defs: {},
})

const instance = client.service('http://my-service.com')

instance.setHeader('my-header', 'my-value')

await instance.call('io.example.doStuff')
```

</td>
<td>

```ts
import { XrpcClient } from '@atproto/xrpc'

const instance = new XrpcClient(
  async (url, init) => {
    const headers = new Headers(init.headers)

    headers.set('my-header', 'my-value')

    // Custom logic here

    const fullUrl = new URL(url, 'http://my-service.com')

    return fetch(fullUrl, { ...init, headers })
  },
  [
    {
      lexicon: 1,
      id: 'io.example.doStuff',
      defs: {},
    },
  ],
)

await instance.call('io.example.doStuff')
```

</td>
</tr>
</table>

> If your fetch handler does not require any "custom logic", and all you need is an `XrpcClient` that makes its HTTP requests towards a static service URL the previous example can be simplified to:

Reviewer 2: ```suggestion
If your fetch handler does not require any "custom logic", and all you need is an `XrpcClient` that makes its HTTP requests towards a static service URL, the previous example can be simplified to:
```
